httpcache.go

```diff
@@ -227,9 +229,22 @@ func (t *Transport) RoundTrip(req *http.Request) (resp *http.Response, err error
 				resp.Header.Set(fakeHeader, reqValue)
 			}
 		}
-		respBytes, err := httputil.DumpResponse(resp, true)
-		if err == nil {
-			t.Cache.Set(cacheKey, respBytes)
+		switch req.Method {
+		case "GET":
+			// Delay caching until EOF is read.
+			resp.Body = newCachedReadCloser(resp.Body, func(r io.ReadCloser) {
```

Comment:
Actually, now that I wrote that up, that doesn't looks so great because the main logic ends up being on top... But I have a better idea. The `newCachedReadCloser` func, by now, offers very little value, aside from the negative value of obscuring the names of parameters. Let's just get rid of it:

```Go
resp.Body = cachedReadCloser{
	r:     resp.Body,
	onEOF: func(r io.ReadCloser) {
		response := *resp
		response.Body = r
		respBytes, err := httputil.DumpResponse(&response, true)
		if err == nil {
			t.Cache.Set(cacheKey, respBytes)
		}
	},
}
```

But to make it more clear that `r` and `onEOF` need to be set, and `buf` is fine to leave as zero value, you should rename export the former and leave latter unexported:

```Go
// cachedReadCloser is a wrapper around ReadCloser R that calls OnEOF
// handler with a full copy of the content read from that ReadCloser
// when EOF is read.
type cachedReadCloser struct {
	R io.ReadCloser
	// OnEOF is called with a copy of the content of R when EOF is read from R.
	OnEOF func(io.ReadCloser)

	buf bytes.Buffer // buf stores a copy of the content of r.
}
```

Since `cachedReadCloser` is not exported, it doesn't make a difference, other than to signify some fields are meant to be set by callers and the rest are internal implementation details of `cachedReadCloser`.